File: solutions/haskell/151-backspace-string-compare.hs
{-
  Problem 151: Backspace String Compare (LeetCode 844)
  Difficulty: Easy
  Language: Haskell
-}
backspaceCompare :: String -> String -> Bool
backspaceCompare s t = build s == build t
  where build = foldl (\acc ch -> if ch == '#' then init' acc else acc ++ [ch]) ""
        init' [] = []
        init' xs = init xs

File: solutions/rust/151-backspace-string-compare.rs
/**
 * Problem 151: Backspace String Compare (LeetCode 844)
 * Difficulty: Easy
 * Language: Rust
 */
fn backspace_compare(s: &str, t: &str) -> bool {
    let build = |input: &str| -> String {
        input.chars().fold(String::new(), |mut acc, ch| {
            if ch == '#' { acc.pop(); } else { acc.push(ch); }
            acc
        })
    };
    build(s) == build(t)
}
